RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,615 @@
+
+    1. Hybrid tensors can be considered as sparse arrays with strided
+       tensors as elements. The `fill_value` attribute represents an
+       unspecified element of the array, hence, the fill value has all
+       the properties that any array element has, including the data
+       type, shape, and memory location.
+
+    2. A non-scalar fill value of a hybrid tensor is more memory-efficient
+       than if the fill value were a scalar.
+
+       For instance, when applying a row-wise function, say, `softmax`
+       to a hybrid tensor, each row may induce a different fill value.
+       As an example, consider a 2D hybrid tensor with one sparse and
+       one dense dimension in COO format, and apply softmax along the
+       first dimension assuming the default fill value `0`. We have:
+
+       ```python
+       >>> A = torch.sparse_coo_tensor(indices=[[0, 3]],
+                                       values=[[.11, .12],
+                                               [.31, .32]],
+                                       size=(4, 2))
+       >>> A.to_dense()
+       tensor([[0.1100, 0.1200],
+               [0.0000, 0.0000],
+               [0.0000, 0.0000],
+               [0.3100, 0.3200]])
+       >>> A.to_dense().softmax(0)
+       tensor([[0.2492, 0.2503],
+               [0.2232, 0.2220],
+               [0.2232, 0.2220],
+               [0.3044, 0.3057]])
+       ```
+
+       Observe that the softmax result can be represented as a sparse
+       tensor with the same sparsity structure (same indices) as the
+       input only when the fill value is a 1-D dense tensor, that is:
+
+       ```python
+       A.softmax(0) == torch.sparse_coo_tensor(indices=A.indices(),
+                                               values=[[0.2492, 0.2503],
+                                                       [0.3044, 0.3057]],
+                                               size=A.size(),
+                                               fill_value=[0.2232, 0.2220])
+       ```
+
+       The alternative, if the fill value was defined as scalar,
+       would be memory inefficient because the result of
+       `A.softmax(0)` is a full sparse tensor with all elements
+       specified.
+
+6.  The fill value that is specified in the `fill_value` argument of
+    the sparse tensor constructors may have different (smaller) shape
+    than the fill value of the hybrid tensor (as defined by point 5).
+
+    The specified fill value (after converting it to `torch.Tensor`
+    instance) can be acquired via `_fill_value()` method.
+
+    For example, the fill value of a (1+1)-D hybrid tensor can be
+    specified as scalar:
+
+    ```python
+    A = torch.sparse_coo_tensor(indices=[[0, 3]],
+                                values=[[.11, .12], [.31, .32]],
+                                size=(4, 2),
+                                fill_value = 1.2)
+    A.fill_value() -> torch.tensor([1.2, 1.2])
+    A._fill_value() -> torch.tensor(1.2)
```

Comment:
can we just return a full-shaped tensor but restrided with strides of 0? Then I don't think you need separate `_fill_value` method.

The element-wise operators can have a fast path for the case when it's really a scalar being broadcasted to the hybrid shape. And general logic would be pretty much the same as for TensorIterator broadcasting rules.

Reply:
One thing that is a little tricky here, is that if you get the restrided tensor with zero and then do an operation on it like `add_(2)`, that will cause the entire tensor to get materialized. Maybe TensorIterator shouldn't do that...